### PR TITLE
Multiline Input v1

### PR DESCRIPTION
**Pull request by libertymayc**
Version 1
a. Variants and styling
b. States (disabled, readonly)
c. Adornments (static, button)
d. Validation status
e. Fixed number of rows
f. Full border

### CODE REVIEW

File: .changeset/green-crabs-sniff.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"@salt-ds/core": minor
+---
+
+Added container div to `StatusAdornment`
```
Comment (joshwooding) [inline .changeset/green-crabs-sniff.md:5]: Why was this change needed?

File: packages/core/src/status-adornment/StatusAdornment.css [hunk 2]
```diff
@@ -1,4 +1,8 @@
 .saltStatusAdornment {
+  height: var(--salt-text-lineHeight);
```
Comment (joshwooding) [inline packages/core/src/status-adornment/StatusAdornment.css:2]: Is this to fix the icon haven't different alignment to text? If so we've already used different techniques that don't affect the component directly e.g. Banner. I spoke to the designers to other day about this design quirk.

File: packages/lab/src/multiline-input/MultilineInput.css [hunk 3]
```diff
@@ -0,0 +1,224 @@
+/* Style applied to the root element */
+.saltMultilineInput {
+  --multilineInput-borderColor: var(--salt-editable-borderColor);
+  --multilineInput-borderStyle: var(--salt-editable-borderStyle);
+  --multilineInput-outlineColor: var(--salt-focused-outlineColor);
+  --multilineInput-borderWidth: var(--salt-size-border);
+  --multilineInput-paddingLeft: var(--salt-spacing-100);
+
+  align-items: center;
+  background: var(--saltMultilineInput-background, var(--multilineInput-background));
```
Comment (joshwooding) [inline packages/lab/src/multiline-input/MultilineInput.css:10]: I probably wouldn't worry adding the salt* variables for now.

File: packages/lab/src/multiline-input/MultilineInput.css [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+  color: var(--saltMultilineInput-color, var(--salt-text-primary-foreground));
+  display: inline-flex;
+  font-family: var(--salt-text-fontFamily);
+  font-size: var(--saltMultilineInput-fontSize, var(--salt-text-fontSize));
+  height: auto;
+  line-height: var(--saltMultilineInput-lineHeight, var(--salt-text-lineHeight));
+  min-height: var(--saltMultilineInput-minHeight, var(--salt-size-base));
+  min-width: var(--saltMultilineInput-minWidth, 4em);
+  padding-left: var(--saltMultilineInput-paddingLeft, var(--multilineInput-paddingLeft));
+  padding-right: var(--saltMultilineInput-paddingRight, var(--salt-spacing-100));
```
Comment (joshwooding) [inline packages/lab/src/multiline-input/MultilineInput.css:21]:
```suggestion
  padding-inline: var(--salt-spacing-100);
```

File: packages/lab/src/multiline-input/MultilineInput.tsx [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+   */
+  emptyReadOnlyMarker?: string;
+  /**
+   * End adornment component
+   */
+  endAdornment?: ReactNode;
+  /**
+   * Styling variant with full border. Defaults to false
+   */
+  fullBorder?: boolean;
```
Comment (joshwooding) [inline packages/lab/src/multiline-input/MultilineInput.tsx:38]: I'm feeling conflicted about having this prop. We've normally tried to push this sort of customisation as patterns

File: packages/lab/src/multiline-input/MultilineInput.tsx [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+  /**
+   * [Attributes](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea#Attributes) applied to the `textarea` element.
+   */
+  textAreaProps?: TextareaHTMLAttributes<HTMLTextAreaElement>;
+  /**
+   * Optional ref for the input component
```
Comment (joshwooding) [inline packages/lab/src/multiline-input/MultilineInput.tsx:44]:
```suggestion
   * Optional ref for the textarea component
```

File: .changeset/green-crabs-sniff.md (hunk 1 above)
Comment (libertymayc) [reply .changeset/green-crabs-sniff.md:5]: I thought since the original class name is now being applied to the outer div

File: packages/lab/src/multiline-input/MultilineInput.tsx (hunk 5 above)
Comment (libertymayc) [reply packages/lab/src/multiline-input/MultilineInput.tsx:38]: cc @bhoppers2008 thoughts?

File: packages/lab/src/multiline-input/MultilineInput.tsx [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+      emptyReadOnlyMarker = "—",
+      endAdornment,
+      fullBorder = false,
+      id,
+      textAreaProps = {},
+      textAreaRef,
+      placeholder,
+      readOnly,
+      role,
+      rows = 3,
```
Comment (feliciackh) [inline packages/lab/src/multiline-input/MultilineInput.tsx:90]: docs indicated the default is 4, not sure which is the correct one 4 or 3

Comment (libertymayc) [reply packages/lab/src/multiline-input/MultilineInput.tsx:90]: good spot thank you! it changed from 4 to 3

File: packages/lab/src/multiline-input/MultilineInput.css [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+  min-height: calc(var(--salt-size-base) + var(--salt-spacing-100));
+  padding-left: var(--salt-spacing-100);
+  column-gap: var(--salt-spacing-100);
+}
+
+/* Style applied if `fullBorder={true}` */
+.saltMultilineInput-fullBorder {
+  --saltMultilineInput-border: var(--salt-size-border) var(--salt-container-borderStyle) var(--multilineInput-borderColor);
+  --saltMultilineInput-paddingLeft: var(--salt-spacing-100);
+  --saltMultilineInput-activationIndicator-border: none;
```
Comment (alycrys) [inline packages/lab/src/multiline-input/MultilineInput.css:203]: i believe you need to override `border`. Currently it's `border: none;`

File: packages/lab/stories/multiline-input/multiline-input.stories.tsx [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    <FlowLayout style={{ width: "366px" }}>
+      <MultilineInput readOnly defaultValue="Value" />
+      <MultilineInput readOnly variant="secondary" defaultValue="Value" />
+      <MultilineInput readOnly fullBorder defaultValue="Value" />
+      <MultilineInput
+        readOnly
+        fullBorder
+        variant="secondary"
+        defaultValue="Value"
+      />
```
Comment (alycrys) [inline packages/lab/stories/multiline-input/multiline-input.stories.tsx:61]: would you need to add `fullBorder` version on `disabled` also

File: packages/lab/stories/multiline-input/multiline-input.stories.tsx [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        fullBorder
+        validationStatus="warning"
+        defaultValue="Value"
+      />
+      <MultilineInput validationStatus="success" defaultValue="Value" />
+      <MultilineInput
+        fullBorder
+        validationStatus="success"
+        defaultValue="Value"
+      />
```
Comment (alycrys) [inline packages/lab/stories/multiline-input/multiline-input.stories.tsx:88]: is there any difference for variant="secondary"?

File: packages/lab/stories/multiline-input/multiline-input.stories.tsx
Comment (alycrys) [inline packages/lab/stories/multiline-input/multiline-input.stories.tsx:1]: I'd also add stories for placeholder and textAlign?

File: packages/lab/src/multiline-input/MultilineInput.tsx [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+   */
+  fullBorder?: boolean;
+  /**
+   * [Attributes](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea#Attributes) applied to the `textarea` element.
+   */
+  textAreaProps?: TextareaHTMLAttributes<HTMLTextAreaElement>;
+  /**
+   * Optional ref for the textarea component
+   */
+  textAreaRef?: Ref<HTMLTextAreaElement>;
```
Comment (alycrys) [inline packages/lab/src/multiline-input/MultilineInput.tsx:46]: I'm inclined to call these `inputProps` and `inputRef` to be inline with the component name. Feels a bit strange

File: packages/lab/src/multiline-input/MultilineInput.tsx [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+      ? emptyReadOnlyMarker
+      : defaultValueProp;
+
+    const {
+      "aria-describedby": textAreaDescribedBy,
+      "aria-labelledby": textAreaLabelledBy,
+      onBlur,
+      onChange,
+      onFocus,
+      required: isRequired,
```
Comment (alycrys) [inline packages/lab/src/multiline-input/MultilineInput.tsx:128]: don't think you need this

File: packages/lab/src/multiline-input/MultilineInput.css (hunk 8 above)
Comment (libertymayc) [reply packages/lab/src/multiline-input/MultilineInput.css:203]: ooh I deleted something on one commit and its broken. fixing now

File: packages/lab/src/multiline-input/MultilineInput.tsx [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+  extends Omit<ComponentPropsWithoutRef<"div">, "defaultValue">,
+    Pick<
+      ComponentPropsWithoutRef<"input">,
+      "disabled" | "value" | "defaultValue" | "placeholder"
+    > {
+  /**
+   * The marker to use in an empty read only Input.
+   * Use `''` to disable this feature. Defaults to '—'.
+   */
+  emptyReadOnlyMarker?: string;
```
Comment (alycrys) [inline packages/lab/src/multiline-input/MultilineInput.tsx:30]: I feel this looks a bit strange on the multiline input. Can we double check with design?

File: packages/lab/src/multiline-input/MultilineInput.tsx [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+   */
+  readOnly?: boolean;
+  /**
+   * Start adornment component
+   */
+  startAdornment?: ReactNode;
+  /**
+   * Alignment of text within container. Defaults to "left"
+   */
+  textAlign?: "left" | "center" | "right";
```
Comment (alycrys) [inline packages/lab/src/multiline-input/MultilineInput.tsx:58]: This seems to align only the textarea content. Do we really need it for textarea? Looks funky

File: packages/lab/stories/multiline-input/multiline-input.stories.tsx (hunk 10 above)
Comment (libertymayc) [reply packages/lab/stories/multiline-input/multiline-input.stories.tsx:129]: it's the same as Input so no difference :)

File: packages/lab/stories/multiline-input/multiline-input.stories.tsx [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+  typeof MultilineInput
+> = (args) => {
+  return (
+    <FlowLayout style={{ width: "366px" }}>
+      <MultilineInput
+        startAdornment={<FlagIcon />}
+        endAdornment={<PinSolidIcon />}
+        validationStatus="success"
+        defaultValue="Value"
+      />
```
Comment (alycrys) [inline packages/lab/stories/multiline-input/multiline-input.stories.tsx:178]: Please check this with rows={1}

File: packages/lab/src/multiline-input/MultilineInput.css [hunk 16]
```diff
@@ -0,0 +1,246 @@
+/* Style applied to the root element */
+.saltMultilineInput {
+  --multilineInput-borderColor: var(--salt-editable-borderColor);
+  --multilineInput-borderStyle: var(--salt-editable-borderStyle);
+  --multilineInput-outlineColor: var(--salt-focused-outlineColor);
+  --multilineInput-borderWidth: var(--salt-size-border);
```
Comment (alycrys) [inline packages/lab/src/multiline-input/MultilineInput.css:6]: this var name is a bit confusing. Name suggests it's the wrapper border, but it's used only for the activationIndicator. Can we rename somehow?

File: packages/lab/stories/multiline-input/multiline-input.stories.tsx [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+            <Text>%</Text>
+            <FilterClearIcon />
+          </>
+        }
+        validationStatus="error"
+        defaultValue="Value"
+        {...args} 
+      />
+    </FlowLayout>
+  );
```
Comment (alycrys) [inline packages/lab/stories/multiline-input/multiline-input.stories.tsx:192]: can we add a story for `readonly` with adornments, please?

File: packages/lab/src/multiline-input/MultilineInput.tsx (hunk 13 above)
Comment (libertymayc) [reply packages/lab/src/multiline-input/MultilineInput.tsx:30]: pending @bhoppers2008

File: packages/lab/src/multiline-input/MultilineInput.tsx (hunk 14 above)
Comment (libertymayc) [reply packages/lab/src/multiline-input/MultilineInput.tsx:58]: pending @bhoppers2008

**Review by joshwooding — CHANGES_REQUESTED** [review]
Can you add tests please

File: packages/lab/src/multiline-input/MultilineInput.tsx (hunk 11 above)
Comment (libertymayc) [reply packages/lab/src/multiline-input/MultilineInput.tsx:46]: thoughts on this? @joshwooding

File: packages/lab/src/__tests__/__e2e__/multiline-input/MultilineInput.cy.tsx [hunk 18]
```diff
@@ -0,0 +1,174 @@
+import { ChangeEvent, useState } from "react";
+import { FormField, FormFieldLabel, Button } from "@salt-ds/core";
+import { MultilineInput } from "@salt-ds/lab";
+
```
Comment (alycrys) [inline packages/lab/src/__tests__/__e2e__/multiline-input/MultilineInput.cy.tsx:4]: let's add the accessibility check also

File: packages/lab/stories/multiline-input/multiline-input.stories.tsx (hunk 15 above)
Comment (alycrys) [reply packages/lab/stories/multiline-input/multiline-input.stories.tsx:212]: I think you can remove it now. Alignment is fixed ;)

File: packages/lab/stories/multiline-input/multiline-input.stories.tsx [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+      />
+      <MultilineInput
+        readOnly
+        startAdornment={<FlagIcon />}
+        endAdornment={
+          <Button disabled>
+            <BookmarkSolidIcon />
+          </Button>
+        }
+        defaultValue="Value"
```
Comment (alycrys) [inline packages/lab/stories/multiline-input/multiline-input.stories.tsx:193]:
```suggestion
        defaultValue="ReadOnly Value"
```

File: packages/lab/stories/multiline-input/multiline-input.stories.tsx [hunk 20]
```diff
@@ -0,0 +1,227 @@
...
+          <Button>
+            <BookmarkSolidIcon />
+          </Button>
+        }
+        defaultValue="Value"
+        {...args}
+      />
+      <MultilineInput
+        readOnly
+        startAdornment={<FlagIcon />}
```
Comment (alycrys) [inline packages/lab/stories/multiline-input/multiline-input.stories.tsx:187]: When using a button as a startAdornment in a readOnly state, the button is clickable. It is not consistent with the endAdornment.

File: packages/lab/src/__tests__/__e2e__/multiline-input/MultilineInput.cy.tsx [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+describe("GIVEN an MultilineInput", () => {
+  it("SHOULD have no a11y violations on load", () => {
+    cy.mount(<MultilineInput defaultValue="The default value" />);
+    cy.checkAxeComponent();
+  });
+
+  describe("WHEN cy.mounted as an uncontrolled component", () => {
```
[the remark this answers is not in the document]
Comment (libertymayc) [reply packages/lab/src/__tests__/__e2e__/multiline-input/MultilineInput.cy.tsx:11]: good idea

File: packages/lab/stories/multiline-input/multiline-input.stories.tsx (hunk 19 above)
Comment (libertymayc) [reply packages/lab/stories/multiline-input/multiline-input.stories.tsx:193]: I think these will change with the guidance coming from content soon to some realistic examples

File: packages/lab/stories/multiline-input/multiline-input.stories.tsx (hunk 20 above)
Comment (libertymayc) [reply packages/lab/stories/multiline-input/multiline-input.stories.tsx:187]: if it's readonly, it's the users responsibility to add disabled prop to the button as well (same in Input)
otherwise, we would have to add an input context for button, to ensure it is actually disabled (can't be done just through styling)

**Review by alycrys — DISMISSED** [review]

File: packages/lab/src/multiline-input/MultilineInput.tsx (hunk 13 above)
Comment (libertymayc) [reply packages/lab/src/multiline-input/MultilineInput.tsx]: Removed this one for now. may come back. @bhoppers2008 needs to look into it

File: packages/lab/src/multiline-input/MultilineInput.tsx (hunk 14 above)
Comment (libertymayc) [reply packages/lab/src/multiline-input/MultilineInput.tsx]: removed; not needed for multiline

File: packages/lab/src/multiline-input/MultilineInput.css [hunk 22]
```diff
@@ -0,0 +1,245 @@
...
+  outline: var(--salt-focused-outlineWidth) var(--salt-focused-outlineStyle) var(--multilineInput-outlineColor);
+}
+
+/* Styling when focused if `disabled={true}` or `readOnly={true}` */
+.saltMultilineInput-focused.saltMultilineInput-readOnly,
+.saltMultilineInput-focused.saltMultilineInput-disabled {
+  --multilineInput-activationIndicator-borderWidth: var(--salt-size-border);
+
+  outline: none;
+}
```
Comment (alycrys) [inline packages/lab/src/multiline-input/MultilineInput.css:137]: this is applicable only when readOnly (no need for disabled). But should the readOnly be focusable (have focused class)?

File: packages/lab/src/multiline-input/MultilineInput.tsx [hunk 23]
```diff
@@ -0,0 +1,194 @@
...
+   */
+  readOnly?: boolean;
+  /**
+   * Start adornment component
+   */
+  startAdornment?: ReactNode;
+  /**
+   * Validation status.
+   */
+  validationStatus?: "error" | "warning" | "success";
```
Comment (alycrys) [inline packages/lab/src/multiline-input/MultilineInput.tsx:53]: there seem to be some inconsistencies on the `validationStatus` type. Maybe we can re-think this when adding the new statuses

File: packages/lab/src/multiline-input/MultilineInput.css [hunk 24]
```diff
@@ -0,0 +1,237 @@
...
+
+/* Style applied to button start adornment if first child */
+.saltMultilineInput-startAdornmentContainer .saltButton:first-child {
+  margin-left: calc(var(--salt-spacing-50) * -1);
+}
+
+/* Style applied to button start adornment if last child */
+.saltMultilineInput-endAdornmentContainer .saltButton:last-child {
+  margin-right: calc(var(--salt-spacing-50) * -1);
+}
```
Comment (alycrys) [inline packages/lab/src/multiline-input/MultilineInput.css:230]: sorry for being a pain, but would you mind moving the adornments styles closer together :P.

File: packages/lab/src/multiline-input/MultilineInput.css [hunk 25]
```diff
@@ -0,0 +1,237 @@
...
+.saltMultilineInput-fullBorder:active,
+.saltMultilineInput-fullBorder.saltMultilineInput-focused {
+  --multilineInput-activationIndicator-borderWidth: var(--salt-editable-borderWidth-active);
+}
+
+/* Styling when focused if `disabled={true}` or `readOnly={true}` when `fullBorder={true}` */
+.saltMultilineInput-fullBorder.saltMultilineInput-readOnly,
+.saltMultilineInput-fullBorder.saltMultilineInput-focused.saltMultilineInput-readOnly,
+.saltMultilineInput-fullBorder.saltMultilineInput-readOnly:hover,
+.saltMultilineInput-fullBorder.saltMultilineInput-focused.saltMultilineInput-disabled,
```
[the remark this answers is not in the document]
Comment (libertymayc) [reply packages/lab/src/multiline-input/MultilineInput.css:209]: oh yes, I fixed it and removed one of them but not this one; will fix

File: packages/lab/src/multiline-input/MultilineInput.css (hunk 24 above)
Comment (libertymayc) [reply packages/lab/src/multiline-input/MultilineInput.css:226]: done

File: packages/lab/src/multiline-input/MultilineInput.tsx (hunk 5 above)
Comment (bhoppers2008) [reply packages/lab/src/multiline-input/MultilineInput.tsx:33]: My preference would be prop. It'll be frequently used (perhaps more so with than without), and with the new multiline input, it's less a customisation and more an explicit design decision.

**Review by alycrys — APPROVED** [review]

**Review by ZarkoPernar — APPROVED** [review]

**Review by alycrys — APPROVED** [review]
